### PR TITLE
feat: DirectX 11 Rendering

### PR DESCRIPTION
**Pull request by ashifolfi**
Initial prep work:
- [x] setup cmake to support swapping render backends and devices via command line
- [x] make adding new renderers and separating renderer specific libraries out eaiser

Rendering:
- [ ] ImGui
- [ ] Shaders
- [ ] Meshes
- [ ] Textures

non conclusive list of things todo. will have things added or removed as needed.

### CODE REVIEW

**Comment by ashifolfi** [comment]
The default rendering device and backend are set to OpenGL no matter the system.

File: engine/render/backend/RenderBackend.h [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 #if defined(CHIRA_USE_GL_41) || defined(CHIRA_USE_GL_43)
```
Comment (craftablescience) [inline engine/render/backend/RenderBackend.h:3]: If you're using `DX11` this should probably be changed to `GL41` and `GL43` for consistency

File: engine/render/backend/api/BackendDX.cpp [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+#include <map>
+#include <stack>
+#include <string>
+
+#include <core/Assertions.h>
+#include <core/Logger.h>
+
+using namespace chira;
+
+CHIRA_CREATE_LOG(DX);
```
Comment (craftablescience) [inline engine/render/backend/api/BackendDX.cpp:14]: DX11 and DX12 are fundamentally different APIs, make this specifically a DX11 backend -- you won't be able to share much between 11 and 12 to my knowledge

File: engine/render/backend/api/CMakeLists.txt [hunk 3]
```diff
@@ -1,5 +1,14 @@
-list(APPEND CHIRA_ENGINE_HEADERS
+if (${CHIRA_RENDER_BACKEND} MATCHES "OpenGL")
+    list(APPEND CHIRA_ENGINE_HEADERS
         ${CMAKE_CURRENT_LIST_DIR}/BackendGL.h)
-
-list(APPEND CHIRA_ENGINE_SOURCES
+    list(APPEND CHIRA_ENGINE_SOURCES
         ${CMAKE_CURRENT_LIST_DIR}/BackendGL.cpp)
+
+elseif (${CHIRA_RENDER_BACKEND} MATCHES "DirectX")
```
Comment (craftablescience) [inline engine/render/backend/api/CMakeLists.txt:7]: Possible values for this variable should be listed in the root cmake

**Review by craftablescience — COMMENTED** [review]
I understand this is a draft PR, but since I don't have access to Discord right now here are some things I noticed

File: CMakeLists.txt [hunk 4]
```diff
@@ -72,14 +76,14 @@ message(STATUS "  CHIRA_TREAT_WARNINGS_AS_ERRORS: ${CHIRA_TREAT_WARNINGS_AS_ERRO
...
 message(STATUS "  CHIRA_USE_STEAMWORKS: ${CHIRA_USE_STEAMWORKS}")
+message(STATUS "  CHIRA_RENDER: \n      BACKEND: ${CHIRA_RENDER_BACKEND}\n      DEVICE: ${CHIRA_RENDER_DEVICE}")
 
 
 # You should add this to your own CMake to load Chira Engine's dependencies in non-Windows environments
 set(CMAKE_SKIP_BUILD_RPATH FALSE)
 set(CMAKE_BUILD_RPATH_USE_ORIGIN TRUE)
 set(CMAKE_INSTALL_RPATH $ORIGIN)
 
-
```
[the remark this answers is not in the document]
Comment (ashifolfi) [reply CMakeLists.txt:82]: I'm actually very unsure how these got removed. I think Visual Studio auto removed them

File: CMakeLists.txt [hunk 5]
```diff
@@ -181,27 +184,16 @@ list(APPEND IMGUI_SOURCES
         ${CMAKE_CURRENT_SOURCE_DIR}/engine/thirdparty/imgui/imgui_tables.cpp
         ${CMAKE_CURRENT_SOURCE_DIR}/engine/thirdparty/imgui/imgui_widgets.cpp)
 
+# include rendering option macros/functions
+include(cmake/rendering.cmake)
```
[the remark this answers is not in the document]
Comment (ashifolfi) [reply CMakeLists.txt:188]: This was done specifically during the development process so I can easily modify the macros and whatnot that are inside of this without having to dig through the main cmakelists everytime.

File: engine/render/backend/RenderBackend.h (hunk 1 above)
Comment (ashifolfi) [reply engine/render/backend/RenderBackend.h:3]: true.

File: engine/render/backend/api/BackendDX.cpp (hunk 2 above)
Comment (ashifolfi) [reply engine/render/backend/api/BackendDX.cpp:14]: :+1: good to know. I hadn't touched the DX12 docs yet. only DX11.

File: engine/render/backend/api/CMakeLists.txt (hunk 3 above)
Comment (ashifolfi) [reply engine/render/backend/api/CMakeLists.txt:7]: will do!

File: CMakeLists.txt [hunk 6]
```diff
@@ -72,14 +76,14 @@ message(STATUS "  CHIRA_TREAT_WARNINGS_AS_ERRORS: ${CHIRA_TREAT_WARNINGS_AS_ERRO
 message(STATUS "  CHIRA_BUILD_HEADLESS: ${CHIRA_BUILD_HEADLESS}")
 message(STATUS "  CHIRA_USE_DISCORD: ${CHIRA_USE_DISCORD}")
 message(STATUS "  CHIRA_USE_STEAMWORKS: ${CHIRA_USE_STEAMWORKS}")
+message(STATUS "  CHIRA_RENDER: \n      BACKEND: ${CHIRA_RENDER_BACKEND}\n      DEVICE: ${CHIRA_RENDER_DEVICE}")
```
[the remark this answers is not in the document]
Comment (ashifolfi) [reply CMakeLists.txt:79]: I can understand this. I'll apply the change

File: cmake/rendering.cmake [hunk 7]
```diff
@@ -0,0 +1,14 @@
+# makes the required changes between renderers much easier to implement via single lists and variables
+
+macro("render_def" def)
```
[the remark this answers is not in the document]
Comment (ashifolfi) [reply cmake/rendering.cmake:3]: mind giving a possible replacement name? I'm unsure what else to name it in that case.

File: cmake/rendering.cmake [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+    list(APPEND CHIRA_ENGINE_DEFINITIONS CHIRA_USE_${def})
+endmacro()
+
+macro("imgui_platform" backend render)
```
[the remark this answers is not in the document]
Comment (ashifolfi) [reply cmake/rendering.cmake:7]: I didn't know if there was any specific reason you separated these into their own include so I kept it like that. the macro was made so I could easily swap out names with less to write.

File: CMakeLists.txt [hunk 9]
```diff
@@ -59,6 +59,10 @@ option(CHIRA_BUILD_HEADLESS "Build Chira Engine without any visible window(s). Z
 option(CHIRA_USE_DISCORD "Build Chira Engine with Discord rich presence features if possible" ON)
 option(CHIRA_USE_STEAMWORKS "Build Chira Engine with Steamworks API features if possible" ON)
 
+# Renderers!
+set(CHIRA_RENDER_BACKEND "OpenGL" CACHE STRING "The rendering backend ChiraEngine will use")
+set(CHIRA_RENDER_DEVICE "OpenGL" CACHE STRING "The rendering device ChiraEngine will use (usually the same as the backend)")
```
[the remark this answers is not in the document]
Comment (ashifolfi) [reply CMakeLists.txt:64]: why is it called `DeviceGL` in that case if it's actually the SDL device?

File: CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashifolfi) [reply CMakeLists.txt:189]: it makes it easier to locate the things I might need to change when I don't have to sift through a ton of code to do so.

File: CMakeLists.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (craftablescience) [reply CMakeLists.txt:64]: `DeviceSDL_GL` was ugly

File: CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (craftablescience) [reply CMakeLists.txt:189]: fair enough

File: cmake/rendering.cmake (hunk 8 above)
[the remark this answers is not in the document]
Comment (craftablescience) [reply cmake/rendering.cmake:7]: honestly i put them in a separate include because i didnt think to use cmake variables at the time

**Comment by craftablescience** [comment]
Added a framework for adding render backends and/or devices in the latest PR

**Comment by craftablescience** [comment]
is this superseded by `feat/d3d11`?

**Comment by craftablescience** [comment]
Yeah this is definitely superseded by `feat/d3d11`, closing

**Pull request closed without merging**